### PR TITLE
Integration test for adding a new StableToken

### PR DESCRIPTION
**Pull request by asaj**
### Description

This PR adds an integration test for adding a new StableToken, mimicking the planned process for deployment of cEUR, and future additional stable assets.

### Other changes

None.

### Tested

These are tests!

### Related issues

None.

### Backwards compatibility

Yes, just tests.

### CODE REVIEW

File: packages/protocol/test/common/integration.ts [hunk 1]
```diff
@@ -559,3 +559,153 @@ contract('Integration: Exchange', (accounts: string[]) => {
...
+contract('Integration: Adding StableToken', (accounts: string[]) => {
+  const Exchange: RegistryContract = artifacts.require('Exchange')
+  const StableToken: StableTokenContract = artifacts.require('StableToken')
+  let exchangeABC: ExchangeInstance
+  let stableTokenABC: StableTokenInstance
+  const sellAmount = 1
+  const decimals = 18
+
+  // 0. Make ourselves the owner of the various contracts we will need to interact with, as
+  // passing a governance proposal for each one will be a pain in the butt.
```
Comment (yorhodes) [inline packages/protocol/test/common/integration.ts:572]: lol
```suggestion
  // passing a governance proposal for each one will be a pain.
```

File: packages/protocol/test/common/integration.ts [hunk 2]
```diff
@@ -559,3 +559,153 @@ contract('Integration: Exchange', (accounts: string[]) => {
     })
   })
 })
+
+contract('Integration: Adding StableToken', (accounts: string[]) => {
```
Comment (yorhodes) [inline packages/protocol/test/common/integration.ts:563]: should we parameterize this like we did above for extra safety?
```javascript
Array.from([
  ['Exchange', 'StableToken'], // USD
  ['ExchangeEUR', 'StableTokenEUR'], // EUR
]).forEach(([exchangeId, stableTokenId]) =>
```

File: packages/protocol/test/common/integration.ts [hunk 3]
```diff
@@ -559,3 +559,153 @@ contract('Integration: Exchange', (accounts: string[]) => {
...
+    const registry = await getDeployedProxiedContract('Registry', artifacts)
+    const contractsToOwn = ['Freezer', 'Registry', 'Reserve', 'SortedOracles']
+    const proposalTransactions = await Promise.all(
+      contractsToOwn.map(async (x: string) => {
+        return {
+          value: 0,
+          destination: (await getDeployedProxiedContract(x, artifacts)).address,
+          data: Buffer.from(
+            stripHexEncoding(
+              // Any contract's `setOwner` function will work here as the function signatures are all the same.
```
Comment (yorhodes) [inline packages/protocol/test/common/integration.ts:591]: 🔥

File: packages/protocol/test/common/integration.ts [hunk 4]
```diff
@@ -559,3 +559,153 @@ contract('Integration: Exchange', (accounts: string[]) => {
...
+    await timeTravel(config.governance.dequeueFrequency, web3)
+    // @ts-ignore
+    const txData = governance.contract.methods.approve(proposalId, dequeuedIndex).encodeABI()
+    await multiSig.submitTransaction(governance.address, 0, txData, {
+      from: accounts[0],
+    })
+    await timeTravel(config.governance.approvalStageDuration, web3)
+    await governance.vote(proposalId, dequeuedIndex, VoteValue.Yes)
+    await timeTravel(config.governance.referendumStageDuration, web3)
+    await governance.execute(proposalId, dequeuedIndex)
```
Comment (yorhodes) [inline packages/protocol/test/common/integration.ts:621]: this is surprisingly clean 👏

File: packages/protocol/test/common/integration.ts [hunk 5]
```diff
@@ -559,3 +559,153 @@ contract('Integration: Exchange', (accounts: string[]) => {
...
+  //   a) Deploy the contracts. For simplicity, omit proxies for now.
+  //   b) Initialize and register the contracts
+  //   c) Confirm mento is effectively frozen
+  describe('When the contracts have been deployed and initialized', () => {
+    before(async () => {
+      exchangeABC = await Exchange.new()
+      stableTokenABC = await StableToken.new()
+
+      // TODO: This is not very readable, what are these parameters and why were these values chosen?
+      // TODO: Which happens first, initialization or registration?
```
Comment (yorhodes) [inline packages/protocol/test/common/integration.ts:634]: looks like we register before initializing
https://github.com/celo-org/celo-monorepo/blob/05ca067f6152302842af190f7a2dc3a84d63a48d/packages/protocol/scripts/truffle/make-release.ts#L154

Comment (yorhodes) [reply packages/protocol/test/common/integration.ts:634]: registration isn't done here?

**Review by yorhodes — COMMENTED** [review]
a few comments

File: packages/protocol/test/common/integration.ts [hunk 6]
```diff
@@ -559,3 +559,153 @@ contract('Integration: Exchange', (accounts: string[]) => {
...
+  const StableToken: StableTokenContract = artifacts.require('StableToken')
+  let exchangeABC: ExchangeInstance
+  let stableTokenABC: StableTokenInstance
+  const sellAmount = 1
+  const decimals = 18
+
+  // 0. Make ourselves the owner of the various contracts we will need to interact with, as
+  // passing a governance proposal for each one will be a pain in the butt.
+  before(async () => {
+    const proposalId = 2
```
Comment (yorhodes) [inline packages/protocol/test/common/integration.ts:574]: why is this 2?

File: packages/protocol/test/common/integration.ts (hunk 5 above)
Comment (asaj) [reply packages/protocol/test/common/integration.ts:634]: Good catch

File: packages/protocol/test/common/integration.ts [hunk 7]
```diff
@@ -559,3 +559,153 @@ contract('Integration: Exchange', (accounts: string[]) => {
...
+        stableToken.address,
+        '5000000000000000000000',
+        '1300000000000000000000',
+        '300',
+        '1'
+      )
+    })
+
+    it(`should be impossible to buy stable token`, async () => {
+      await goldToken.approve(exchange.address, 1)
```
Comment (yorhodes) [inline packages/protocol/test/common/integration.ts:657]: `goldToken` or `stableTokenAbc`

Comment (yorhodes) [inline packages/protocol/test/common/integration.ts:657]:
```suggestion
      await stableTokenAbc.approve(exchange.address, 1)
```

File: packages/protocol/test/common/integration.ts [hunk 8]
```diff
@@ -559,3 +559,153 @@ contract('Integration: Exchange', (accounts: string[]) => {
...
+  //   c) Confirm mento is functional
+  describe('When the contracts have been unfrozen and the mento has been activated', () => {
+    before(async () => {
+      await reserve.addToken(stableToken.address)
+      await freezer.unfreeze(stableTokenAbc.address)
+      await freezer.unfreeze(exchangeAbc.address)
+    })
+
+    it(`should be possible to buy stable token`, async () => {
+      await goldToken.approve(exchange.address, 1)
```
Comment (yorhodes) [inline packages/protocol/test/common/integration.ts:702]:
```suggestion
      await stableTokenAbc.approve(exchange.address, 1)
```

File: packages/protocol/test/common/integration.ts [hunk 9]
```diff
@@ -559,3 +559,153 @@ contract('Integration: Exchange', (accounts: string[]) => {
...
+    })
+  })
+
+  // 3. Mimic the state of the world post-mento-activation-proposal
+  //   a) Add the stable token to the reserve
+  //   b) Unfreeze the mento
+  //   c) Confirm mento is functional
+  describe('When the contracts have been unfrozen and the mento has been activated', () => {
+    before(async () => {
+      await reserve.addToken(stableToken.address)
```
Comment (yorhodes) [inline packages/protocol/test/common/integration.ts:696]:
```suggestion
      await reserve.addToken(stableTokenAbc.address)
```

File: packages/protocol/test/common/integration.ts [hunk 10]
```diff
@@ -559,3 +559,153 @@ contract('Integration: Exchange', (accounts: string[]) => {
     })
   })
 })
+
+contract('Integration: Adding StableToken', (accounts: string[]) => {
+  const Exchange: RegistryContract = artifacts.require('Exchange')
+  const StableToken: StableTokenContract = artifacts.require('StableToken')
+  let exchangeABC: ExchangeInstance
+  let stableTokenABC: StableTokenInstance
```
Comment (yorhodes) [inline packages/protocol/test/common/integration.ts:567]:
```suggestion
  let exchangeAbc: ExchangeInstance
  let stableTokenAbc: StableTokenInstance
```

File: packages/protocol/test/common/integration.ts [hunk 11]
```diff
@@ -559,3 +559,153 @@ contract('Integration: Exchange', (accounts: string[]) => {
...
+        '0x000000000000000000000000000000000000ce10',
+        '1000000000000000000000000',
+        '47304000',
+        [accounts[0]],
+        [web3.utils.toWei(1, 'ether')],
+        'ExchangeABC'
+      )
+      await exchange.initialize(
+        '0x000000000000000000000000000000000000ce10',
+        stableToken.address,
```
Comment (yorhodes) [inline packages/protocol/test/common/integration.ts:648]:
```suggestion
        stableTokenAbc.address,
```

File: packages/protocol/test/common/integration.ts [hunk 12]
```diff
@@ -559,3 +559,153 @@ contract('Integration: Exchange', (accounts: string[]) => {
...
+  describe('When the contracts have been unfrozen and the mento has been activated', () => {
+    before(async () => {
+      await reserve.addToken(stableToken.address)
+      await freezer.unfreeze(stableTokenAbc.address)
+      await freezer.unfreeze(exchangeAbc.address)
+    })
+
+    it(`should be possible to buy stable token`, async () => {
+      await goldToken.approve(exchange.address, 1)
+      await exchange.buy(1, 0, true)
```
Comment (yorhodes) [inline packages/protocol/test/common/integration.ts:703]:
```suggestion
      await goldToken.approve(exchangeAbc.address, 1)
      await exchangeAbc.buy(1, 0, true)
```

**Comment by asaj** [comment]
Dupe of https://github.com/celo-org/celo-monorepo/pull/7179

**Pull request closed without merging**